File: frontend/cte/lib/app/data/model/cte_ferroviario_ferrovia_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/infra/infra_imports.dart';
import 'package:cte/app/data/model/model_imports.dart';
import 'package:cte/app/data/domain/domain_imports.dart';

class CteFerroviarioFerroviaModel {
	int? id;
	int? idCteFerroviario;
	String? cnpj;
	String? codigoInterno;
	String? ie;
	String? nome;
	String? logradouro;
	String? numero;
	String? complemento;
	String? bairro;
	int? codigoMunicipio;
	String? nomeMunicipio;
	String? uf;
	String? cep;
	CteFerroviarioModel? cteFerroviarioModel;

	CteFerroviarioFerroviaModel({
		this.id,
		this.idCteFerroviario,
		this.cnpj,
		this.codigoInterno,
		this.ie,
		this.nome,
		this.logradouro,
		this.numero,
		this.complemento,
		this.bairro,
		this.codigoMunicipio,
		this.nomeMunicipio,
		this.uf,
		this.cep,
		this.cteFerroviarioModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cnpj',
		'codigo_interno',
		'ie',
		'nome',
		'logradouro',
		'numero',
		'complemento',
		'bairro',
		'codigo_municipio',
		'nome_municipio',
		'uf',
		'cep',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cnpj',
		'Codigo Interno',
		'Ie',
		'Nome',
		'Logradouro',
		'Numero',
		'Complemento',
		'Bairro',
		'Codigo Municipio',
		'Nome Municipio',
		'Uf',
		'Cep',
	];

	CteFerroviarioFerroviaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteFerroviario = jsonData['idCteFerroviario'];
		cnpj = jsonData['cnpj'];
		codigoInterno = jsonData['codigoInterno'];
		ie = jsonData['ie'];
		nome = jsonData['nome'];
		logradouro = jsonData['logradouro'];
		numero = jsonData['numero'];
		complemento = jsonData['complemento'];
		bairro = jsonData['bairro'];
		codigoMunicipio = jsonData['codigoMunicipio'];
		nomeMunicipio = jsonData['nomeMunicipio'];
		uf = CteFerroviarioFerroviaDomain.getUf(jsonData['uf']);
		cep = jsonData['cep'];
		cteFerroviarioModel = jsonData['cteFerroviarioModel'] == null ? CteFerroviarioModel() : CteFerroviarioModel.fromJson(jsonData['cteFerroviarioModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteFerroviario'] = idCteFerroviario != 0 ? idCteFerroviario : null;
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['codigoInterno'] = codigoInterno;
		jsonData['ie'] = ie;
		jsonData['nome'] = nome;
		jsonData['logradouro'] = logradouro;
		jsonData['numero'] = numero;
		jsonData['complemento'] = complemento;
		jsonData['bairro'] = bairro;
		jsonData['codigoMunicipio'] = codigoMunicipio;
		jsonData['nomeMunicipio'] = nomeMunicipio;
		jsonData['uf'] = CteFerroviarioFerroviaDomain.setUf(uf);
		jsonData['cep'] = Util.removeMask(cep);
		jsonData['cteFerroviarioModel'] = cteFerroviarioModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteFerroviario = plutoRow.cells['idCteFerroviario']?.value;
		cnpj = plutoRow.cells['cnpj']?.value;
		codigoInterno = plutoRow.cells['codigoInterno']?.value;
		ie = plutoRow.cells['ie']?.value;
		nome = plutoRow.cells['nome']?.value;
		logradouro = plutoRow.cells['logradouro']?.value;
		numero = plutoRow.cells['numero']?.value;
		complemento = plutoRow.cells['complemento']?.value;
		bairro = plutoRow.cells['bairro']?.value;
		codigoMunicipio = plutoRow.cells['codigoMunicipio']?.value;
		nomeMunicipio = plutoRow.cells['nomeMunicipio']?.value;
		uf = plutoRow.cells['uf']?.value != '' ? plutoRow.cells['uf']?.value : 'AC';
		cep = plutoRow.cells['cep']?.value;
		cteFerroviarioModel = CteFerroviarioModel();
		cteFerroviarioModel?.fluxo = plutoRow.cells['cteFerroviarioModel']?.value;
	}	

	CteFerroviarioFerroviaModel clone() {
		return CteFerroviarioFerroviaModel(
			id: id,
			idCteFerroviario: idCteFerroviario,
			cnpj: cnpj,
			codigoInterno: codigoInterno,
			ie: ie,
			nome: nome,
			logradouro: logradouro,
			numero: numero,
			complemento: complemento,
			bairro: bairro,
			codigoMunicipio: codigoMunicipio,
			nomeMunicipio: nomeMunicipio,
			uf: uf,
			cep: cep,
		);			
	}

	
}

File: backend/cte/src/model/cte_ferroviario_ferrovia_model.py
from src import db
from src.model.cte_ferroviario_model import CteFerroviarioModel


class CteFerroviarioFerroviaModel(db.Model):
    __tablename__ = 'cte_ferroviario_ferrovia'

    id = db.Column(db.Integer, primary_key=True)
    cnpj = db.Column(db.String(14))
    codigo_interno = db.Column(db.String(10))
    ie = db.Column(db.String(20))
    nome = db.Column(db.String(60))
    logradouro = db.Column(db.String(250))
    numero = db.Column(db.String(60))
    complemento = db.Column(db.String(60))
    bairro = db.Column(db.String(60))
    codigo_municipio = db.Column(db.Integer)
    nome_municipio = db.Column(db.String(60))
    uf = db.Column(db.String(2))
    cep = db.Column(db.String(8))
    id_cte_ferroviario = db.Column(db.Integer, db.ForeignKey('cte_ferroviario.id'))

    cte_ferroviario_model = db.relationship('CteFerroviarioModel', foreign_keys=[id_cte_ferroviario])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_ferroviario = data.get('idCteFerroviario')
        self.cnpj = data.get('cnpj')
        self.codigo_interno = data.get('codigoInterno')
        self.ie = data.get('ie')
        self.nome = data.get('nome')
        self.logradouro = data.get('logradouro')
        self.numero = data.get('numero')
        self.complemento = data.get('complemento')
        self.bairro = data.get('bairro')
        self.codigo_municipio = data.get('codigoMunicipio')
        self.nome_municipio = data.get('nomeMunicipio')
        self.uf = data.get('uf')
        self.cep = data.get('cep')

    def serialize(self):
        return {
            'id': self.id,
            'idCteFerroviario': self.id_cte_ferroviario,
            'cnpj': self.cnpj,
            'codigoInterno': self.codigo_interno,
            'ie': self.ie,
            'nome': self.nome,
            'logradouro': self.logradouro,
            'numero': self.numero,
            'complemento': self.complemento,
            'bairro': self.bairro,
            'codigoMunicipio': self.codigo_municipio,
            'nomeMunicipio': self.nome_municipio,
            'uf': self.uf,
            'cep': self.cep,
            'cteFerroviarioModel': self.cte_ferroviario_model.serialize() if self.cte_ferroviario_model else None,
        }